### PR TITLE
Check alloc() respects condition "charged >= refunded"

### LINKED ISSUE #2024: Check `alloc` method in `Ext` always charges more than refunds
### Problem to Solve

.

### Possible Solution

.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by ark0f**
Resolves #2024.

### CODE REVIEW

File: core-processor/src/ext.rs [hunk 1]
```diff
@@ -883,42 +883,74 @@ impl Ext {
...
+            fn new(ext: &'a mut Ext) -> Self {
+                Self { charged: 0, ext }
+            }
+
+            fn calculate_gas(&self, alloc: u32, mem_grow: u32) -> u64 {
+                let alloc = alloc as u64;
+                let mem_grow = mem_grow as u64;
+                0_u64
+                    .saturating_add(alloc.saturating_mul(self.ext.context.config.alloc_cost))
+                    .saturating_add(mem_grow.saturating_mul(self.ext.context.config.mem_grow_cost))
```
Comment (SabaunT) [inline core-processor/src/ext.rs:901]:
```suggestion
                    alloc
                    .saturating_mul(self.ext.context.config.alloc_cost)
                    .saturating_add(mem_grow.saturating_mul(self.ext.context.config.mem_grow_cost))
```
is something like that possible/better?

File: core-processor/src/ext.rs [hunk 2]
```diff
@@ -883,42 +883,74 @@ impl Ext {
         pages: WasmPageNumber,
         mem: &mut impl Memory,
     ) -> Result<WasmPageNumber, ProcessorError> {
+        struct AllocGasCharger<'a> {
```
Comment (SabaunT) [inline core-processor/src/ext.rs:886]: seems this is too noisy being inside the function, isn't it?

**Review by SabaunT — CHANGES_REQUESTED** [review]
I'd expect something like a fuzz test or a type level enforcement of the condition "charged >= refunded"

File: core-processor/src/ext.rs (hunk 1 above)
Comment (ark0f) [reply core-processor/src/ext.rs:901]: It's possible but I doubt it better as it has lower readability

File: core-processor/src/ext.rs [hunk 3]
```diff
@@ -198,6 +198,42 @@ impl AsTerminationReason for ProcessorError {
     }
 }
 
+struct ChargedAllocGas {
+    amount: u64,
+}
+
+impl ChargedAllocGas {
+    fn calculate_gas(ext: &Ext, alloc: u32, mem_grow: u32) -> u64 {
```
Comment (breathx) [inline core-processor/src/ext.rs:206]: impl Into<u64>

File: core-processor/src/ext.rs [hunk 4]
```diff
@@ -604,13 +640,14 @@ impl EnvExt for Ext {
         self.charge_gas_runtime(RuntimeCosts::Free)?;
 
         let result = self.context.allocations_context.free(page);
+        self.return_and_store_err(result)?;
 
         // Returns back gas for allocated page if it's new
-        if result.is_ok() && !self.context.allocations_context.is_init_page(page) {
+        if !self.context.allocations_context.is_init_page(page) {
             self.refund_gas(self.context.pages_config.alloc_cost)?;
```
Comment (breathx) [inline core-processor/src/ext.rs:647]: return and store err?

File: core-processor/src/ext.rs [hunk 5]
```diff
@@ -198,6 +198,42 @@ impl AsTerminationReason for ProcessorError {
     }
 }
 
+struct ChargedAllocGas {
```
Comment (breathx) [inline core-processor/src/ext.rs:201]: some comments would be also good

File: core-processor/src/ext.rs [hunk 6]
```diff
@@ -908,42 +945,23 @@ impl Ext {
...
-            self.context
-                .pages_config
-                .alloc_cost
-                .saturating_mul((pages.raw() - new_allocated_pages_num) as u64),
-        );
+            .map(|page| !self.context.allocations_context.is_init_page(page) as u32)
+            .sum();
 
-        self.refund_gas(gas_to_return_back)?;
+        let not_allocated = pages.raw() - new_allocated_pages_num;
```
Comment (breathx) [inline core-processor/src/ext.rs:962]: saturating?

Comment (breathx) [reply core-processor/src/ext.rs:962]: or comment on guarantees

File: core-processor/src/ext.rs [hunk 7]
```diff
@@ -1171,4 +1195,130 @@ mod tests {
...
+                self.0 = self.0.add(pages).map_err(|_| MemoryError::OutOfBounds)?;
+                Ok(())
+            }
+
+            fn size(&self) -> WasmPageNumber {
+                self.0
+            }
+
+            fn write(&mut self, _offset: u32, _buffer: &[u8]) -> Result<(), MemoryError> {
+                unreachable!()
```
Comment (breathx) [inline core-processor/src/ext.rs:1223]: unimplemented

**Review by breathx — APPROVED** [review]

**Review by SabaunT — APPROVED** [review]

**Pull request merged**